### ISSUE TITLE
What is supposed to be the strongest DIB line in the spectral range?

### ISSUE THREAD

**Issue by davidwhogg** [jeffreymei/StellarSpectroscopy#77]
Let's work on a close-up of that.  We can save memory and use all the data if we just work on a small spectral range.

**Comment by jeffreymei** [comment]
I found a paper here that describes DIBS -
http://iopscience.iop.org/0004-637X/727/1/33 - and it looks like 6284A is
our best bet. Though, I found this catalogue (
http://ipag.obs.ujf-grenoble.fr/~desertf/Bibliography/Jenniskens_1994_AAS_106_39.pdf)
which has many DIBS  in our spectral range and there are many (>10) that
have high FWHM values so we have many options.

2015-02-23 15:06 GMT+04:00 David W. Hogg <notifications@github.com>:

**Comment by davidwhogg** [comment]
Now, can you make a zoom in on our three-panel plots at 6284?  Like from 5984 to 6584 ?

Also, is your code set up so that if we make a smaller spectral range (like the one above) you don't use nearly as much memory and we can do the full data set?

**Comment by jeffreymei** [comment]
To do that I would really just have to truncate the wl array used in the
continuum estimation/flux normalization. I'm not sure how much time is
spent on that though (as opposed to storing this data when I call all the
spectra). The change is easy to make though and I can see if it helps much.

2015-03-02 17:22 GMT+04:00 David W. Hogg <notifications@github.com>:

**Comment by davidwhogg** [comment]
My point is that you could then run on *all* of the spectra without running out of memory.

**Comment by jeffreymei** [comment]
I tried it but it doesn't seem really useful, it shaves a minute off a 8-minute process. I'll try running with a bigger sample to see if the savings are different. Here's the plot, the blue line in the middle is the DIB: https://files.nyu.edu/mym226/public/reduced_fit_sample.png

**Comment by jeffreymei** [comment]
yeah, half the data set took 53 minutes which is more or less the same as before. I mean, I don't think it's a big problem to take ~2 hours to do each fitting.

**Comment by davidwhogg** [comment]
I thought the point was that you run out of memory, not time?  Am I wrong?

Also, if you work with far fewer wavelengths, it should take far less time.  Is all the time spent in "getting the data"?  If so, we can speed up your code *enormously*.  Can you profile your code using cProfile?

Also, can I see the results as you increase the data set to the full data set?

**Comment by jeffreymei** [comment]
The memory thing is fine now, the old memory issue was because I don't know how many subspectra are in each fits file and so I can't pre-define an array so I had to use a list, which was bad for memory. I overcame this by making one long numpy array with a running sum, adding the subspectra for each fits file cumulatively. 

I am profiling now, will let you know how it goes.

**Comment by jeffreymei** [comment]
Here is the fit: https://files.nyu.edu/mym226/public/reduced_fit_sample.png. Blue line is the DIB.

**Comment by jeffreymei** [comment]
Here is the full fit: https://files.nyu.edu/mym226/public/reduced_wl_fit.png. Blue line is the DIB.

**Comment by davidwhogg** [comment]
Okay great!  This is promising -- there is a peak in the right place!

Let's do the same for the next-best DIB candidate.  This is very promising...

**Comment by jeffreymei** [comment]
I am doing a cProfile and following the instructions here: http://ymichael.com/2014/03/08/profiling-python-with-cprofile.html but I can't get cprofilev to work so I'm not sure how to read the output. I used 
$ python -m cProfile -o profile_output new_fit.py
to generate the profile_output file, which is here: https://files.nyu.edu/mym226/public/profile_output if you can make sense of it. I tried opening it in vim but it wasn't very clear to read.

**Comment by davidwhogg** [comment]
try this http://pymotw.com/2/profile/

or try to use the run() command in cProfile -- I think it produces human-readable output.

Or run again but without the -o command and output file -- what does it write to terminal?

**Comment by jeffreymei** [comment]
Here is what happens when I call cont_est (which estimates the continuum and returns the inverse sqrts and wavelength grid) once: https://files.nyu.edu/mym226/public/profile_output_contest.txt I'm not sure why it calls so much.

Do you know how to profile.run something that is not a function? Right now the fitting code is under an if init == name thing so I can't import it into another code to try and use profile.run().

**Comment by jeffreymei** [comment]
Oh I figured it out. here it is, I truncated the bottom half with the cumulative time of that call being less than 1 second. https://files.nyu.edu/mym226/public/profile_output_newfit.txt It is dominated by the cont_est call and also a lot of numpy.median/mean ones too? Though I don't see why those would be timeconsuming.

**Comment by davidwhogg** [comment]
Does it make sense that np.median is being called 20179945 times?  Can you account for that number, at least roughly?  It sounds large to me.  Oh and why is cont_est called 703 times?
